backend/infrahub/core/diff/query/merge.py

```diff
@@ -289,43 +308,80 @@ async def query_init(self, db: InfrahubDatabase, **kwargs: Any) -> None:
-    WITH attribute_prop_diff
-    MATCH (n:Node {uuid: attribute_prop_diff.node_uuid})-[:HAS_ATTRIBUTE]->(a:Attribute {name: attribute_prop_diff.attribute_name})
-    UNWIND attribute_prop_diff.properties AS property_diff
+    WITH attr_rel_prop_diff
+    CALL {
+        WITH attr_rel_prop_diff
+        OPTIONAL MATCH (n:Node {uuid: attr_rel_prop_diff.node_uuid})
+            -[has_attr:HAS_ATTRIBUTE]
+            ->(attr:Attribute {name: attr_rel_prop_diff.attribute_name})
+        WHERE attr_rel_prop_diff.attribute_name IS NOT NULL
+        AND has_attr.branch IN [$source_branch, $target_branch]
+        RETURN attr
+        ORDER BY has_attr.from DESC
+        LIMIT 1
+    }
+    CALL {
+        WITH attr_rel_prop_diff
+        OPTIONAL MATCH (n:Node {uuid: attr_rel_prop_diff.node_uuid})
+            -[r1:IS_RELATED {branch: $source_branch}]
+            -(rel:Relationship {name: attr_rel_prop_diff.relationship_id})
+            -[r2:IS_RELATED {branch: $source_branch}]
+            -(:Node {uuid: attr_rel_prop_diff.peer_uuid})
+        WHERE attr_rel_prop_diff.relationship_id IS NOT NULL
+        RETURN rel
+        ORDER BY r1.from DESC, r2.from DESC
+        LIMIT 1
+    }
+    WITH attr_rel_prop_diff, COALESCE(attr, rel) AS attr_rel
+    UNWIND attr_rel_prop_diff.properties AS property_diff
     // ------------------------------
-    // handle updates for properties under this attribute
+    // handle updates for properties under this attribute/relationship
     // ------------------------------
     CALL {
-        WITH a, property_diff
+        WITH attr_rel, property_diff
         // ------------------------------
         // identify the correct property node to link
         // ------------------------------
         CALL {
-            WITH a, property_diff
+            WITH attr_rel, property_diff
             OPTIONAL MATCH (peer:Node {uuid: property_diff.value})
             WHERE property_diff.property_type IN ["HAS_SOURCE", "HAS_OWNER"]
-            OPTIONAL MATCH (bool:Boolean {value: property_diff.value})
-            WHERE property_diff.property_type IN ["IS_VISIBLE", "IS_PROTECTED"]
+            // ------------------------------
+            // the serialized diff might not include the values for IS_VISIBLE and IS_PROTECTED in
+            // some cases, so we need to figure them out here
+            // ------------------------------
+            CALL {
+                WITH attr_rel, property_diff
+                OPTIONAL MATCH (attr_rel)-[r_vis_pro]->(bool:Boolean)
+                WHERE property_diff.property_type IN ["IS_VISIBLE", "IS_PROTECTED"]
+                AND r_vis_pro.branch IN [$source_branch, $target_branch]
+                AND type(r_vis_pro) = property_diff.property_type
+                AND (property_diff.value IS NULL OR bool.value = property_diff.value)
+                RETURN bool
+                ORDER BY r_vis_pro.from DESC
+                LIMIT 1
+            }
```

Comment:
if we don't have a specific value for IS_VISIBLE or IS_PROTECTED, we will get the latest value on the source branch and use that